### DISCUSSION TITLE
Prisma + PostgreSQL: Prefixing uuid-primary keys with dbgenerated()

### DISCUSSION THREAD

**Discussion by felixhagspiel** [prisma/prisma#12243] [Q&A] [answered]
I would like to prefix my uuids. I am using prisma + postgresql. 

Example:

User: `usr_3df6bbd4-de24-465c-9080-1b3673d1d749`
Merchants: `mer_e05ab1f9-c5f4-404c-b71c-8a623ea0941f`

... and so on.

As this is currently not supported by default, I am planning to do it like this:

```
    model User {
      id                         String                @id @default(dbgenerated("(concat('usr_', gen_random_uuid()))::TEXT"))
    }
```

This results in this SQL:

```
    -- CreateTable
    CREATE TABLE "User" (
        "id" TEXT NOT NULL DEFAULT (concat('usr_', gen_random_uuid()))::TEXT,

        CONSTRAINT "User_pkey" PRIMARY KEY ("id")
    );
```

I have tested it and it works, but will I run into any problems later on with performance or relations with this approach? Or is this the right way to do it? Thank you!

**Comment by austincrim** [comment, marked as answer]
Hey @felixhagspiel 👋 ,

This looks like a great use of `dbgenerated`! I don't foresee any apparent performance issues with the concatenation of UUID generation happening the the DB itself. The `id` will be typed as a plain `String` as usual, but relations should all operate the same.

If you hit anything unexpected, please open an [issue](https://github.com/prisma/prisma/issues/new?assignees=&labels=kind%2Fbug&template=bug_report.yml) with more details.

**Reply by felixhagspiel** [reply]
Hey @austincrim thanks for replying! Would there be any reasons regarding to performance (or others) to use the native `UUID` type of postgres instead of  prismas  `uuid()` / `TEXT` column? Or is there no (noticeable) difference? I remember reading some years ago that is somewhat a bad practice to use `TEXT` as a primary key type in postgres

**Comment by liesislukas** [comment]
would be cool if `cuid` and `uuid` prisma's functions would allow to provide prefix. Would not need to check database's extensions & would just work out of the box. Prisma docs could even suggest to add unique prefixes to IDs. Debug time is reduced heavily in many cases when you just have some ID in logs/support request etc and that id has prefix which says exactly what object is it. `@default(cuid("u-"))` would be generating ids like `u-34wrsdg...` 

update. ran some tests, postgres 14 does not require any extensions for this to work, but still `cuid("u-")` is what i would prefer 

see my post here for more details:

https://dba.stackexchange.com/a/315347/103306

**Reply by austincrim** [reply]
Hey @liesislukas 👋,

Feel free to 👍 [this existing feature request](https://github.com/prisma/prisma/issues/6719) or add your specific use case if it hasn't been covered in that thread already.

**Reply by 0xdevalias** [reply]
See also:

- https://github.com/prisma/prisma/issues/3391

**Comment by illusion77** [comment]
Thanks! It can be even shorter after ignoring type inference.
``` sql
id            String         @id @default(dbgenerated("'user-' || gen_random_uuid()"))
```

**Comment by prageeth** [comment]
I believe it's better to do this in code instead of the schema itself. I was using this method in my own projects and I kept copy/pasting it all the time, so I ended up creating an open source plugin: https://www.npmjs.com/package/prisma-prefixed-ids
